lib/dry/cli/utils/files/file_system.rb

```diff
@@ -0,0 +1,190 @@
+          #   require "dry/cli/utils/files/file_system"
+          #
+          #   fs = Dry::CLI::Utils::Files::FileSystem.new
+          #   fs.mkdir("/usr/var/project/file.rb")
+          #   # creates all the directory structure (/usr/var/project)
+          #   # where file.rb will eventually live
+          #
+          # @since x.x.x
+          # @api private
+          def mkdir_p(path, **kwargs)
+            mkdir(
+              file.dirname(path), **kwargs
+            )
+          end
+
+          # Removes (deletes) a file
+          #
+          # @see https://ruby-doc.org/stdlib/libdoc/fileutils/rdoc/FileUtils.html#method-c-rm
+          #
+          # @param path [String] the file to remove
+          #
+          # @since x.x.x
+          # @api private
+          def rm(path, **kwargs)
+            file_utils.rm(path, **kwargs)
+          end
+
+          # Removes (deletes) a directory
+          #
+          # @see https://ruby-doc.org/stdlib/libdoc/fileutils/rdoc/FileUtils.html#method-c-remove_entry_secure
+          #
+          # @param path [String] the directory to remove
+          #
+          # @since x.x.x
+          # @api private
+          def rm_rf(path, *args)
+            file_utils.remove_entry_secure(path, *args)
+          end
+
+          # Reads the entire file specified by name as individual lines,
+          # and returns those lines in an array
+          #
+          # @see https://ruby-doc.org/core/IO.html#method-c-readlines
+          #
+          # @param path [String] the file to read
+          #
+          # @since x.x.x
+          # @api private
+          def readlines(path, *args)
+            file.readlines(path, *args)
+          end
+
+          # Check if the given file exist.
+          #
+          # @see https://ruby-doc.org/core/File.html#method-c-exist-3F
+          #
+          # @param path [String] the file to check
+          #
+          # @since x.x.x
+          # @api private
+          def exist?(path)
+            file.exist?(path)
+          end
+
+          # Check if the given path is a directory.
+          #
+          # @see https://ruby-doc.org/core/File.html#method-c-directory-3F
+          #
+          # @param path [String] the directory to check
+          #
+          # @since x.x.x
+          # @api private
+          def directory?(path)
+            file.directory?(path)
+          end
+
+          # Opens (or creates) a new file for read/write operations.
+          #
+          # @see https://ruby-doc.org/core/File.html#method-c-open
+          #
+          # @param path [String] the target file
+          #
+          # @since x.x.x
+          # @api private
+          def open(path, *args, &blk)
+            file.open(path, *args, &blk)
+          end
+
+          private
+
+          # @since x.x.x
+          # @api private
+          attr_reader :file
+
+          # @since x.x.x
+          # @api private
+          attr_reader :file_utils
```

Comment:
No biggie but in general in dry/rom we're defining readers at the top and we never-ever define them as private, because it makes debugging horrible and also method visibility can always be bypassed in Ruby, and people do that, so we can mark things as private but that won't prevent misusage. In general my belief is that documenting API as private is enough. Another nice benefit is that when you change your mind, it's just a documentation update, no code changes needed.

I *know* people have completely opposite opinions though so it'd be good to find consensus here. Again, no biggie but let's address this at some point 🙂

Reply:
@solnic I don't agree here, but consistency across libs is top priority. Fixed in https://github.com/dry-rb/dry-cli/pull/112/commits/14be7d3f9466bde94458bc16bd600b39b220f568